File: Backend/interfaces/api/v1/vehicle.py
"""
interfaces/api/v1/vehicle.py
-----------------------------
REST endpoints for the ``Vehicle`` resource.

Endpoints:
  POST  /vehicles          — create a vehicle
  GET   /vehicles          — list vehicles (paginated)
  GET   /vehicles/{id}     — get a single vehicle
"""
from __future__ import annotations

import uuid

from fastapi import APIRouter, Depends, Query

from application.vehicle.commands.create_vehicle import CreateVehicleCommand
from application.vehicle.commands.create_vehicle_handler import CreateVehicleHandler
from application.vehicle.queries.get_vehicle import GetVehicleQuery
from application.vehicle.queries.get_vehicle_handler import GetVehicleHandler
from application.vehicle.queries.list_vehicles import ListVehiclesQuery
from application.vehicle.queries.list_vehicles_handler import ListVehiclesHandler
from application.vehicle.vehicle_dto import CreateVehicleRequest
from interfaces.api.deps import (
    get_create_vehicle_handler,
    get_get_vehicle_handler,
    get_list_vehicles_handler,
)
from interfaces.api.response import success_response
from interfaces.api.v1.mappers import to_vehicle_response

router = APIRouter(prefix="/vehicles", tags=["vehicles"])


@router.post("")
def create_vehicle(
    payload: CreateVehicleRequest,
    handler: CreateVehicleHandler = Depends(get_create_vehicle_handler),
):
    vehicle = handler.handle(
        CreateVehicleCommand(
            plate_number=payload.plate_number,
            manufacturer=payload.manufacturer,
            model=payload.model,
            vehicle_image_url=payload.vehicle_image_url,
            color=payload.color,
            production_year=payload.production_year,
            vin=payload.vin,
        )
    )
    return success_response(data=to_vehicle_response(vehicle).model_dump(by_alias=True))


@router.get("/{vehicle_id}")
def get_vehicle(
    vehicle_id: uuid.UUID,
    handler: GetVehicleHandler = Depends(get_get_vehicle_handler),
):
    vehicle = handler.handle(GetVehicleQuery(vehicle_id=vehicle_id))
    return success_response(data=to_vehicle_response(vehicle).model_dump(by_alias=True))


@router.get("")
def list_vehicles(
    limit: int = Query(default=20, ge=1, le=100),
    handler: ListVehiclesHandler = Depends(get_list_vehicles_handler),
):
    vehicles = handler.handle(ListVehiclesQuery(limit=limit))
    return success_response(
        data=[to_vehicle_response(v).model_dump(by_alias=True) for v in vehicles]
    )

File: Frontend/src/types/vehicle.ts
export interface VehicleApiDto {
  _id: string;
  plate_number: string;
  manufacturer: string;
  model: string;
  vehicle_image_url: string | null;
  color: string | null;
  production_year: number | null;
  vin: string | null;
  device_id: string | null;
  _created_at: string | null;
  _updated_at: string | null;
}

export interface Vehicle {
  id: string;
  plateNumber: string;
  manufacturer: string;
  model: string;
  vehicleImageUrl: string | null;
  color: string | null;
  productionYear: number | null;
  vin: string | null;
  deviceId: string | null;
  createdAt: string | null;
  updatedAt: string | null;
}

export function mapVehicleFromApi(dto: VehicleApiDto): Vehicle {
  return {
    id: dto._id,
    plateNumber: dto.plate_number,
    manufacturer: dto.manufacturer,
    model: dto.model,
    vehicleImageUrl: dto.vehicle_image_url,
    color: dto.color,
    productionYear: dto.production_year,
    vin: dto.vin,
    deviceId: dto.device_id || null,
    createdAt: dto._created_at,
    updatedAt: dto._updated_at,
  };
}
